### ISSUE TITLE
Authentification Needed

### ISSUE THREAD

**Issue by Lemmings23456** [gcobb321/icloud3#151]
Hi, 
I cannot get iCloud3 back to work.
When I restart HA ( and delete the files with my username in .storage/iCloud/session  and  .storage/iCloud there is a notify in HA, that I should authenticate my iCloud account. But it takes more than 5-10 Minutes before I can enter it and then it is wrong. After that, no more notify is visible. 

`2021-04-29 15:13:56 ERROR (SyncWorker_6) [custom_components.icloud3.device_tracker] Authentication needed for Account (Status Code 450)
Traceback (most recent call last):
  File "/config/custom_components/icloud3/device_tracker.py", line 6776, in _setup_tracked_devices_for_fmf
    friend_valid_emails_msg = self._get_me_device_id(api_friends, friend_valid_emails_msg)
  File "/config/custom_components/icloud3/device_tracker.py", line 6878, in _get_me_device_id
    famshr_devices_data = self.api.devices.response["content"]
  File "/config/custom_components/icloud3/pyicloud_ic3.py", line 660, in devices
    data = FindMyiPhoneServiceManager(service_root, self.session, self.params, self.with_family)
  File "/config/custom_components/icloud3/pyicloud_ic3.py", line 749, in __init__
    self.refresh_client(device_id)
  File "/config/custom_components/icloud3/pyicloud_ic3.py", line 759, in refresh_client
    req = self.session.post(
  File "/usr/local/lib/python3.8/site-packages/requests/sessions.py", line 590, in post
    return self.request('POST', url, data=data, json=json, **kwargs)
  File "/config/custom_components/icloud3/pyicloud_ic3.py", line 173, in request
    return self.request(method, url, **kwargs)
  File "/config/custom_components/icloud3/pyicloud_ic3.py", line 175, in request
    self._raise_error(response.status_code, response.reason)
  File "/config/custom_components/icloud3/pyicloud_ic3.py", line 235, in _raise_error
    raise api_error
custom_components.icloud3.pyicloud_ic3.PyiCloudAPIResponseException: Authentication needed for Account (Status Code 450)`

While starting up, HA shows:
`2021-04-28 15:33:47 WARNING (MainThread) [homeassistant.bootstrap] Waiting on integrations to complete setup: device_tracker`

This alone repeats 5 times and takes 5 minutes. I don't know if icloud3 is the cause.

When I restart iCloud3 there is no more notification inside HA to enter my credentials for iCloud.

configuration.yaml:
device_tracker:
  # ios App ist die Home Assistant App
  - platform: icloud3
    username: !secret icloud_user
    password: !secret icloud_pw
    tracking_method: famshr
    track_devices:
      - ollis_iphone > olli.jpg, Olli
      - iphone_von_julia > julia.jpg, Julia
      - iphone_von_astrid > astrid.jpg, Astrid
      - iphone_marc > marc.png, Marc
      - ipad_von_monika_2 > mutter.png, Mutter
      - iphone_von_ann_marie > annmarie.jpg, Ann-Marie
    inzone_interval: '30 min'
    max_interval: '4 hrs'
    center_in_zone: true
    stationary_inzone_interval: '15 min'
    stationary_still_time: '8 min'
    distance_method: calc
    unit_of_measurement: km
    time_format: 24

config_ic3.yaml:

# --Track Devices Parameters-------------------------------
devices:
 - device_name: ollis_iphone
   name: Olli
   email: ********@****
   picture: olli.jpg
   noiosapp: true 
 - device_name: iphone_von_astrid
   name: Astrid
   email: ********@****
   picture: astrid.jpg
   noiosapp: true 
 - device_name: iphone_marc
   name: Marc
   picture: marc.png 
   noiosapp: true  
 - device_name: iphone_von_ann_marie
   name: Ann-Marie
   picture: annmarie.jpg  
   noiosapp: true  
 - device_name: iphone_von_julia
   name: Julia
   picture: julia.jpg   
   noiosapp: true  
 - device_name: ipad_von_monika_2
   name: Monika
   picture: mutter.png        
   noiosapp: true
# Other devices parameters not normally used

#    iosapp_suffix: _app
#    iosapp_entity: gary_iphone_app
#    track_from_zone: warehouse

#event_log_card_directory: www/custom_cards

# --Zone/Tracking Parameters--------------------------------
inzone_interval: 60 min
max_interval: 4 hrs
center_in_zone: true
stationary_inzone_interval: 30 min
stationary_still_time: 8 min
stationary_zone_offset: 1, 0
travel_time_factor: .6
distance_method: calc

# --Display Parameters--------------------------------------
#display_zone_fname: True
unit_of_measurement: km
time_format: 24

**Comment by gcobb321** [comment]
Don't delete the files yourself. That will not trigger a reauthentication code request. Use Event Log > Actions > Reset iCloud Interface.

**Comment by Lemmings23456** [comment]
Okay,
But I can‘t figure out in what order I have to do the verification to get it to work. It is days now and this happens all the time:

![5E80F88D-4DDC-4722-9E7B-EE3C8571A5DC](https://user-images.githubusercontent.com/59805667/116741048-1be56700-a9f6-11eb-9c3d-97ab5705516e.jpeg)

And now the circle begins again. No notification is visible where I can put the number.

**Comment by gcobb321** [comment]
See the iCloud3 docs [here](https://gcobb321.github.io/icloud3/#/chapters/1.2-getting-started?id=authenticating-your-icloud-account) about the entry of the verification code.

**Comment by Lemmings23456** [comment]
Well thanks but pointing out to the docs doesn't work, that's if everything is fine. But it isn't. I'm trying for days now to get it back to work.
There is a notification where I can enter a code. I entered but nothing happens. The window stays without any message. I can only close the code-enter-window with the x. After that, icloud3 tells me that a verification is required. 
What shall I do now ?

<img width="802" alt="Bildschirmfoto 2021-05-02 um 09 14 21" src="https://user-images.githubusercontent.com/59805667/116805445-f4da8280-ab26-11eb-9c54-ada6fc465f9d.png">

I don't want to use to "iOS App" - maybe the name is some kind of irritating. Don't you mean the Home-Assistant App ? I have tons of iOS Apps on my iPhone. 

There is a parameter "noiosapp" - confusing. It should be sth like "use HA app" true / false. 

`2021-05-02 08:42:56 ERROR (SyncWorker_0) [custom_components.icloud3.device_tracker]  ^3^iCloud3 Error > There was an error logging into the Apple/iCloud Account when iCloud3 was started. Verification can not be completed. Review the iCloud3 Event Log for more information.
2021-05-02 08:42:58 ERROR (SyncWorker_11) [custom_components.icloud3.device_tracker]  ^3^iCloud3 Error > There was an error logging into the Apple/iCloud Account when iCloud3 was started. Verification can not be completed. Review the iCloud3 Event Log for more information.
2021-05-02 08:43:00 ERROR (SyncWorker_3) [custom_components.icloud3.device_tracker]  ^3^iCloud3 Error > There was an error logging into the Apple/iCloud Account when iCloud3 was started. Verification can not be completed. Review the iCloud3 Event Log for more information.
2021-05-02 08:43:01 ERROR (SyncWorker_14) [custom_components.icloud3.device_tracker]  ^3^iCloud3 Error > There was an error logging into the Apple/iCloud Account when iCloud3 was started. Verification can not be completed. Review the iCloud3 Event Log for more information.
2021-05-02 08:43:03 ERROR (SyncWorker_12) [custom_components.icloud3.device_tracker]  ^3^iCloud3 Error > There was an error logging into the Apple/iCloud Account when iCloud3 was started. Verification can not be completed. Review the iCloud3 Event Log for more information.
2021-05-02 08:43:03 ERROR (SyncWorker_19) [custom_components.icloud3.device_tracker]  ^3^iCloud3 Error > There was an error logging into the Apple/iCloud Account when iCloud3 was started. Verification can not be completed. Review the iCloud3 Event Log for more information.
2021-05-02 08:43:04 ERROR (SyncWorker_4) [custom_components.icloud3.device_tracker]  ^3^iCloud3 Error > There was an error logging into the Apple/iCloud Account when iCloud3 was started. Verification can not be completed. Review the iCloud3 Event Log for more information.
2021-05-02 08:43:12 ERROR (SyncWorker_3) [custom_components.icloud3.device_tracker]  ^3^iCloud3 Error > There was an error logging into the Apple/iCloud Account when iCloud3 was started. Verification can not be completed. Review the iCloud3 Event Log for more information.
2021-05-02 08:43:14 ERROR (SyncWorker_13) [custom_components.icloud3.device_tracker]  ^3^iCloud3 Error > There was an error logging into the Apple/iCloud Account when iCloud3 was started. Verification can not be completed. Review the iCloud3 Event Log for more information.
2021-05-02 08:43:14 ERROR (SyncWorker_16) [custom_components.icloud3.device_tracker]  ^3^iCloud3 Error > There was an error logging into the Apple/iCloud Account when iCloud3 was started. Verification can not be completed. Review the iCloud3 Event Log for more information.
2021-05-02 08:43:32 ERROR (SyncWorker_19) [custom_components.icloud3.device_tracker]  ^3^iCloud3 Error > There was an error logging into the Apple/iCloud Account when iCloud3 was started. Verification can not be completed. Review the iCloud3 Event Log for more information.
2021-05-02 08:43:32 ERROR (SyncWorker_0) [custom_components.icloud3.device_tracker]  ^3^iCloud3 Error > There was an error logging into the Apple/iCloud Account when iCloud3 was started. Verification can not be completed. Review the iCloud3 Event Log for more information.`

**Comment by gcobb321** [comment]
If the code entry window does not close, there is some type of communication problem between HA and Apple iCloud servers that verify the code. Some things to try:
1. To see the raw transactions being sent to Apple and returned from them, add the following to your configuration.yaml file:
```
logger:
  logs:
    custom_components.icloud3.pyicloud_ic3.py: debug
```
 The entries will be added to the HA log file. That might show some type of error code. 
 
2. I can’t remember if you had selected to not allow access t to your iCloud account from the alert window the Apple displays when someone tries to access your account or if it was someone else. If it was you, Apple might be blocking your HA system from a accessing it and the authentication process. Log into your account and see if there are any errors or notifications that need to be cleared. 

3. Create another iCloud  account and see if you can authenticate that one. If you can, that would indicate something is going on on Apple’s side with your regular account. 

4. Comment out the iCloud3 platform parameters in your HA config file so it doesn’t load. Install the HA iCloud integration and see if that authenticates. If not, it might generate another Apple error that shows what is going on since you are now trying to access your account with another program. Then disablement delete the iCloud HA integration and try iCloud3 again. 

There is no easy fix when trying to solve Interface problems since everything that happens is on Apple’s  side and out of our control. On my system, I reset the iCloud Interface on the Event Log, get the code window, enter the code and it’s accepted without any problems. That’s why I’m suggesting looking at your account, setting up another account or trying the iCloud integration.

**Comment by pauly7300** [comment]
Hi Gary, this issue seems like it may be at the source of the issue i've messaged you about in the forum [here](https://community.home-assistant.io/t/icloud3-v2-4-an-advanced-device-tracker-custom-component-that-is-integrated-with-the-ios-app-location-monitor/267554/257).

Additionally, i've tried all of the suggestions just above including trying multiple icloud accounts as well as commenting out IC3 and trying the HA iCloud integration.  While it's possible it could be related to my icloud account it seems a bit unlikely since the same behavior is happening to 2 different icloud accounts.  

Additionally i've spun up multiple VM's to try and isolate all other factors and have even downgraded Core to 2021.7.x which is where I had it running last but even that doesn't seem to work. My assumption was the the version of Core was potentially the issue but now I'm even more confused.

**Comment by charliebeals** [comment]
Wanted to put in here that I'm experiencing the same issue:


```
4:04:34p | iCloud3  Error > iCloud Account Authentication needed > Will use iOS App  Tracking Method until the iCloud account is authentication is complete.  See the HA Notification area to continue. iCloud3 will then be  restarted.
-- | --
4:04:34p | iCloud3 Error > An error occurred authenticating c******@a****.com.The  iCloud Web Services may be down or the Username/Password may be  invalid. The iCloud Location Services tracking method is disabled and  the iOS App tracking method will be used. The authentication will be retried in 15 min at 4:19:34.
```

Based on documentation, I would expect this error every 15 minutes until the verification code is entered in the notification window. However, no notification ever shows up:
![image](https://user-images.githubusercontent.com/63005214/150602990-512e45a3-e31f-4ff4-b46c-dc0a46386591.png)

Confirmed the username and password via web browser login. Also generated a OTP via iCloud Settings on my iPhone and appended it to the password in configuration.yaml, then restarted iCloud3, all within about 10 seconds to ensure the OTP didn't expire. No bueno on all counts.

EDIT: Also confirmed every 15 minutes, this error message is repeated:

`iCloud3 Error > An error occurred authenticating c*****@a****.com.The  iCloud Web Services may be down or the Username/Password may be  invalid. The iCloud Location Services tracking method is disabled and  the iOS App tracking method will be used. The authentication will be retried in 15 min at 4:49:40.`

Testing another iCloud account as well, no change.

**Comment by pauly7300** [comment]
Since I replied in this thread earlier, I wanted to mention what ultimately fixed my issue. I must have spent 30 or 40 hours troubleshooting this issue and in the end it was nothing I would have guessed. I had periodically noticed messages in the logs about IPV6 which I typically had disabled in my Network.  Somewhere along the way I had noticed an IPv6 setting on the System tab under Host. It was enabled in Home Assistant. When I disabled it the problem went away. It may not be the same issue identified in this thread but it sounded similar to what I was experiencing so I just wanted to share. 

![image](https://user-images.githubusercontent.com/11494052/150607978-add948fd-9072-48d1-8fd3-8ae06192ae7c.png)

**Comment by gcobb321** [comment]
@pauly7300  That’s strange about the IPv6 setting causing this. Perhaps there is a different encoding used by IPv6 than IPv4 that iCloud was not recognizing. The OTP password is typically used by apps, not programs like HA and iCloud3. iCloud3 needs the plain username/password that you use to log into iCloud or iTunes on a web browser. The access info/security info is stored in cookie files in the HA `config/core/.storage/icloud` directory. To reset them and trigger an immediate request for the 6-digit code, use Event Log > Actions > Reset iCloud Interface.

**Comment by charliebeals** [comment]
@pauly7300 Thanks for the update. I made this same change and it's working now, so it does seem to be the case that IPv6 in some way interferes with the iCloud Login. Based on the error iCloud3 was giving, there wasn't anything at all about the OTP, but rather a bad username/password. ICloud would only grant an OTP once the credentials check out, but I wasn't getting the OTP/login notification on any of my apple devices, making me suppose that the login itself was failing.

**Comment by pauly7300** [comment]
@charliebeals You have no idea how great it is to hear that this corrected for you too.  I went back in my emails and found a sort of summary note I sent to Gary back in October.

TL;DR
In the end I don't think it was a problem with IC3 directly at all and was definitely NOT an issue with credentials.  Something related to that IPV6 setting caused a ~10 min bootstrap timeout/delay which I could never isolate.  That delay preceded and prevented the HA notification service from firing the verification code notification thus preventing me from entering the verification code until after it had expired.  I spent an inordinant amount of time sifting through logs and modifying logging settings to try and find the exact cause of the interaction of that IPV6 setting and the bootstrap timeout behavior but in the end never figured it out.  

Maybe this helps someone else or maybe an HA Dev will spot this and might be able to shed some light.

> Don't ask me how but disabling IPV6 in the 'Host > IP Address' setting screen was all it took.  I'm now starting the back investigation to see when this setting was either added or how/if it was ever set to enabled.  I know beyond a shadow of a doubt that I did NOT enable it and didn't change any other settings in my network 7 weeks ago when I started seeing what I assumed were IC3 issues.
> 
> I do believe there's still something worth investigating generally although it may be more related to HA itself than IC3.  I'll hit the dev forums and let you know what I find out.  See below:  The tell-tale sign of the problem is this bootstrap delay seen in the log snip below.  Right at startup, ic3/pyicloud would call Apple with my creds.  This would generate the native Apple prompt for someone trying to login to the account, and then generating the 6 digit code.  This would sit on the screen...  Meanwhile there's a 300 second bootstrap timeout followed by an additional 300 seconds (of something that I couldn't figure out - might have been in IC3) until finally the configurator notification would fire so I could attempt to enter the verification code.  By then the code had timed out and no way to generate it again without restarting (which obviously repeats the same issue).  
> 
> As it turns out, the installation type that I've used for years includes Supervisor and Host settings as seen in the screen below.  If you recall I mentioned that I had gotten it working in a development environment a couple of weeks back.  That installation didn't include this setting for managing IPV6 as I assume uses the computers native network stack.  I also installed on an RPI this AM as you had suggested and even on that brand new installation, I saw the same exact behavior (IPV6 enabled).  But sure enough, disabling IPV6 on the RPI solved it on that build as well as on each of my other VM installations.  So all in all over the past 7 weeks, I've done at least 5-6 fresh installs across different VM's and RPIs.  I've restarted/rebooted ~150 times, trying different settings each time.  I've changed iCloud PW, tried a completely new separate iCloud account, changed idevice names to rule that out.  Learned some Python and tinkered with Pyicloud manually (cool to finally learn that). ...And countless other things that I can't even remember now.  ha ha.


![unnamed](https://user-images.githubusercontent.com/11494052/150641322-c29a3a4f-8819-4c79-89cf-4909a82de930.png)

**Comment by pauly7300** [comment]
> To reset them and trigger an immediate request for the 6-digit code, use Event Log > Actions > Reset iCloud Interface.

This instruction is completely correct @gcobb321 Gary however i'm not sure if you ever saw the behavior first hand.  When this bootstrap delay is occuring, the Event Log card is blank and you're not able to take the action to reset the iCloud interface.  Here's an older note I had captured while trying to troubleshoot. And again, I believe the blank IC3 Event card behavior is indirectly caused by these bootstrap timeout logs seen in the other post.

> Also, while all this hanging is going on, the ic3 event log card is just blank and when I go to it, these lines show up in the logs.

> 2021-09-08 11:59:39 DEBUG (MainThread) [homeassistant.components.websocket_api.http.connection] [140341136220320] Received {'type': 'call_service', 'domain': 'system_log', 'service': 'write', 'service_data': {'logger': 'frontend.js.latest.202108300', 'message': 'http://192.168.1.177:8123/local/custom_cards/icloud3-event-log-card.js:1481:38 Uncaught TypeError: Cannot convert undefined or null to object'}, 'id': 37}
> 2021-09-08 11:59:39 DEBUG (MainThread) [homeassistant.core] Bus:Handling <Event call_service[L]: domain=system_log, service=write, service_data=logger=frontend.js.latest.202108300, message=http://192.168.1.177:8123/local/custom_cards/icloud3-event-log-card.js:1481:38 Uncaught TypeError: Cannot convert undefined or null to object>
> 2021-09-08 11:59:39 ERROR (MainThread) [frontend.js.latest.202108300] http://192.168.1.177:8123/local/custom_cards/icloud3-event-log-card.js:1481:38 Uncaught TypeError: Cannot convert undefined or null to object
